libraries/chain/fork_database.cpp

```diff
@@ -589,13 +569,17 @@ namespace eosio::chain {
    }
 
    template<class bsp>
-   bsp fork_database<bsp>::get_block(const block_id_type& id) const {
+   bsp fork_database<bsp>::get_block(const block_id_type& id, bool check_root /* = false */) const {
       std::shared_lock g( my->mtx );
-      return my->get_block_impl(id);
+      return my->get_block_impl(id, check_root);
    }
 
    template<class bsp>
-   bsp fork_database_impl<bsp>::get_block_impl(const block_id_type& id) const {
+   bsp fork_database_impl<bsp>::get_block_impl(const block_id_type& id, bool check_root /* = false */) const {
+      if( check_root && root->id() == id ) {
+         return root;
```

Comment:
This is not valid. `root` can just contain `block_header` when it is read from disk. I guess we will see how good of test coverage we have. Should hopefully get some test failures on this PR.

Reply:
> root can just contain block_header when it is read from disk.

Is there a reason that the root is different? Why not store the full block_state? Also I see we do `head = root` so maybe it is the same for `head`? Also the type for `head` and `root` in fork_db is `block_state`, not `block_header_state`.

Also we call `fork_db::reset()` with a `block_state` even though it takes a `block_header_state`. This is the kind of issues which make it difficult to not have `block_state` inherit from `block_header_state`.

I think clearly this PR will not cause any issue, since the two calls to `fork_db::get_block_header()`  did not (and could not) use any member that is not part of `block_header_state`. I think it would be better if fork_db uniforly stored `block_state` structs and not one or the other depending on where you look.

Reply:
I don't disagree, but we have to support the legacy format which does store only the `block_header_state_legacy`.

Reply:
I think the issue would be if someone passes `true` to `check_root` when they are expecting a full `block_state`.

Reply:
> I think the issue would be if someone passes `true` to `check_root` when they are expecting a full `block_state`.

Yes, someone doing that would quickly realize they don't get what they expect. The usage is so limited though (just in create_block_token_*) that I wonder if this is a big issue.